text/3579-pki.md

```diff
@@ -0,0 +1,309 @@
+
+This adds additional questions for future signing and PKI work to be done within the project; in the case of implementing TUF, sigstore, etc. we will have to make additional considerations for how this CA may play into those systems and how it can be included or excluded.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+In addition to signing index files, we could also sign crate files themselves. The signed index files will include a hash of the crate file, which will enable verification, but we should evaluate if there are any circumstances where direct crate file signatures would help.
+
+### Foregoing a Root Certificate Authority
+
+Other frameworks for providing additional security we might want within crates.io, such as TUF, implement other methodologies, such as multi-signer schemes. We consider this a separate problem space, and this initial certificate authority can be used independently or integrated with such a solutions. We do not dismiss this approach, and future plans may involve leveraging our CA for providing an additional level of security to such a solution (Rust signing of delegate keys which are used for signing repositories, for example). However, other uses for a CA exist which cannot directly leverage the multi-signer schema for security. The approach proposed in this RFC bridges from a quorum scheme to a CA model in order to allow the use of widespread existing implementations based on the CA model.
+
+### The Update Framework
+
+This RFC does not preclude deploying TUF as a solution for crate and mirror signing. However, one thing to keep in mind is TUF does not solve issues around the quorum, how it is managed, where signing occurs, or how it occurs; TUF provides a model hierarchy for signing and how trust among keys for a file index occurs, and easily rotatable keylists for hierarchical trust. We would still have the problems around how we select a qourum, how and where signing occurs, and where the trust lies. We would still need to determine how we select our TUF quorum, who they are, how and where they perform signing. All those problems are deferred but TUF - but iterated in this RFC.
+
+Part of this approach is meeting our requirements in a model which is understood and applicable by a broader audiance - doing this for a CA is well understood and implemented, while within TUF is only understood and applicable in that context.
+
+In a future where crates.io implements TUF, this CA and quorum would exist as the Root Role, where then we delegate Target roles to crates.io delegate key, and then utilize the TUF model underneath that. We'd still need to solve our quorum, ceremonies, etc. but it would only apply to TUF - a problem space addressed in this RFC. Said another way, a majority of this RFC applies to problems that we also need to solve in TUF, but at a broader scale utilizing a full CA and not strictly within the TUF model.
```

Comment:
TUF has built-in first-class support for k-of-n threshold signing for establishing a quorum which can be used for role, including the root, and which can be validated end-to-end by clients.

This gives it greater deployment flexibility around how the quorum could operate, e.g. instead of an HSM storing a key-to-the-kingdom and enforcing the quorum around key usage via trusted hardware, it could instead support a quorum of signers who independently store and manage keys using e.g. PIN-protected offline Yubikeys/YubiHSMs/SoloKeys/Nitrokeys in tandem with secure endpoints like a Chromebook.

Achieving quorum for signed updates could simply be making PRs to flat files managed through e.g. GitHub, where someone opens a PR that performs an update, and then k-of-n signers compute a signature over the new contents and add their signatures to the PR. When k-of-n have been achieved, the update can be merged and deployed via CDN.

See the [Python Software Foundation's TUF runbook](https://github.com/psf/psf-tuf-runbook) for a real-world example and working Rust code for managing a TUF root role using this sort of process, which was used to provision the PyPI root role.

Reply:
@tarcieri FWIW, I am hugely in favor of the idea of trying to eliminate the bastion server in favor of just having individual quorum members perform the signing operations locally and submitting the results, if we can arrange to make that work.

To me, the big concern is the degree to which we *expose* the result of that to individual users. "There's a root key for Rust which signs things" is a very well-understood model. "There are N quorum members whose individual signatures must be on anything that's trusted" is a less well-understood model, if we actually expose that to users as what the client is doing to figure out if something is trusted.

If we can *abstract that away*, to where we have N quorum members individually signing things but *to clients doing verification* we can make that look like there's a single root key that signs things, I'm all for it. For instance, if we could embed *that* logic in an HSM or similar, such that the HSM signs things if and only if they're threshold-signed like that, and then others who rely on us can *just* look at the HSM signature, that seems perfectly reasonable.

Reply:
> To me, the big concern is the degree to which we expose the result of that to individual users. "There's a root key for Rust which signs things" is a very well-understood model. "There are N quorum members whose individual signatures must be on anything that's trusted" is a less well-understood model, if we actually expose that to users as what the client is doing to figure out if something is trusted. That seems even more true if the quorum members are rotated.

I'll be using TUF as an example here, as it's the implementation I'm familiar with.

Quorum would not be exposed directly to users, unless they want to dig deeper into how the signatures are made (at which point they'd need to learn about the quorum described in this RFC anyway). In TUF, all the root keys would be included in a single JSON file called the "root role". This file contains the public keys of all trusted keys, plus signatures of at least the quorum of keys.

If a rotation needs to happen, tooling or users would not go and swap a few individual keys. They'd replace the old JSON root role as a whole with a new JSON root role containing the new set of trusted keys. If it's a routine rotation the new root role would need to be signed by a quorum of the keys of the previous role.

From the user's point of view it would be a single file providing the root of trust, like the CA key described in the RFC. ~~If we're worried about confusing users, we could just PEM-encode the TUF root role.~~

Reply:
As an example, [here is the TUF root role for sigstore](https://github.com/sigstore/root-signing/blob/main/repository/repository/root.json). It's a single JSON file, even though it contains 5 quorum keys, 4 signatures, and 2 delegated keys.

Reply:
Threshold EdDSA is far simpler than threshold ECDSA, but anyways threshold signatures are not "accountable" in the sense that you do not know who signed the bad signature.  Accountable sounds pretty important when defending against supplychain attacks.

There are good reasons why organizations might use threshold signatures for their own internal opsec, even when distributing software, like hiding the number of signers from the signers themselves, but accountable sounds more important here, and in the general TUF usage.